### PR TITLE
[distributed] feat: add DeepSpeed ZeRO integration

### PR DESCRIPTION
**Pull request by johndpope**
- Add deepspeed extra to pyproject.toml (deepspeed>=0.15.0)
- Add DeepSpeedConfig dataclass with zero_stage, offload config, NVMe support
- Add ds_checkpointer: save/load via DeepSpeed engine
- Add deepspeed_init: build_ds_config() and init_deepspeed_engine()
- Add deepspeed mode to parallel_state: skip DeviceMesh, use WORLD group
- Add deepspeed mode to clip_grad_norm: no-op (DS handles internally)
- Add deepspeed mode to torch_parallelize: skip FSDP/DDP wrapping
- Update base trainer: DS engine init before backward, engine.step
- Update text trainer: DS engine.step path

### What does this PR do?

> Concise overview of the change. Reference related issues/PRs.

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> Validation results (training curves, eval metrics) for changes not covered by CI.

### API and Usage Example

> Show API changes and usage examples if applicable.

### Design & Code Changes

> High-level design description and specific change list.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

**Comment by Luosuu** [comment]
I do not think deepseed has any advantage against FSDP2 nowadays?

**Comment by johndpope** [comment]
Screenshot 2026-05-18 at 5 54 54 PM

**Comment by Luosuu** [comment]
not sure whether this is necessary considering the maintenance effort

**Comment by FoolPlayer** [comment]
Hi thanks for the contribution.

So far we use FSDP as the training backend, could you clarify the motivation adding DeepSpeed? In particular, what concrete use case is not covered by FSDP2, and do you have benchmarks or failure cases showing where DeepSpeed is needed?

Also, since this would introduce a second distributed backend with separate checkpointing, trainer logic, and compatibility constraints, what is the proposed long-term maintenance plan for keeping it working across trainers, models, SP/EP, and future VeOmni changes?

**Comment by johndpope** [comment]
im attempting to boot up qwen 3.6 on a consumer environment / workstation in an attempt to retrain the attention head of open-dllm to use diffusion - the authors succesfully achieved this with qwen 2.5 on smaller scale - they used veomni  to help bootstrap their trainer - but it's impossible to boot qwen 3.6 without huge amounts of ram / vram. 

the only chance I found as of the other day is to leverage microsoft deepseed which can offload to nvme drive. 
https://arxiv.org/abs/2408.10013 (<- original author from nvidia - guess they quickly sunset this direction) 


deepseed can leverage nvme
Screenshot From 2026-05-20 19-15-27


 https://github.com/scrya-com/Open-dLLM
 i spent today investigating nvfp4 training - and  rearchitecting things around qlora - and stumbled across this framework from nvidia 
https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/9c5dcea6e8233be4033d10fcc552880ddf32687a/src/megatron/bridge/recipes/qwen_vl/qwen35_vl.py#L253


once i can validate the training - then i can confidently estimate costs to train at large possibly with some crowdfunding (from the paper - they claim 4x speed ups in tokens per second. my rough calculations overshot that 3000 / sec)
https://www.reddit.com/r/LocalLLaMA/comments/1tee5ms/can_a_5090_with_qwen36_achieve_3000_toks_bring/

most recent direction to use qlora here 
https://github.com/scrya-com/Open-dLLM/tree/persistent-sparse-adam


i will probably rebuild my implementation of open-dllm around megatron - i can close this branch - but if you see the value by all means cherry pick code.

**Comment by Coach257** [comment]
Maybe it would make sense to build an independent deepspeed_trainer.py and move related modules (e.g., parallel_state, DeepSpeed configs, etc.) into this trainer to keep the integration self-contained.

If you manage to build a DLLM based on Qwen3.6, it would also be helpful to provide a document and an example training script demonstrating how to use this deepspeed_trainer. Otherwise, using a simple Qwen3 training example would also be totally fine.

**Comment by johndpope** [comment]
the code for deepseed is self contained with my repo - it doesnt need upstreaming here. 
i think i get around it for now. 

with some wrangling - I managed to get it training with 4 layers of 64 from qwen 3.6 on 5090 
- it's hitting 400 tokens / second - this may regress when i add more layers. idk - 
https://wandb.ai/snoozie/open-dllm/runs/q4fk9bi9

to get it running - you need to precompute the anchors - as per readme
https://github.com/scrya-com/Open-dLLM (**DEVELOP** BRANCH)  

when i have something more polished - i'll let you know.

File: veomni/checkpoint/ds_checkpointer.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    def save(
+        self,
+        path: str,
+        state: Dict[str, Any],
+        save_async: Optional[bool] = None,
+        **kwargs,
+    ):
+        engine = state["model"]  # In DS mode, model IS the engine
+        tag = os.path.basename(path)  # e.g. "global_step_100"
+        os.makedirs(path, exist_ok=True)
```
Comment (ZihaoW123) [inline veomni/checkpoint/ds_checkpointer.py]: DeepSpeedEngine.save_checkpoint() internally creates all necessary directories. Delete the os.makedirs
